Sources/DocCDocumentation/DocCSymbolInformation.swift

```diff
@@ -0,0 +1,133 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import IndexStoreDB
+package import SemanticIndex
+@_spi(LinkCompletion) @preconcurrency import SwiftDocC
+import SymbolKit
+
+package struct DocCSymbolInformation {
+  let components: [(name: String, information: LinkCompletionTools.SymbolInformation)]
+
+  /// Find the DocCSymbolLink for a given symbol USR.
+  ///
+  /// - Parameters:
+  ///   - usr: The symbol USR to find in the index.
+  ///   - index: The CheckedIndex to search within.
+  package init?(fromUSR usr: String, in index: CheckedIndex) {
+    guard let topLevelSymbolOccurrence = index.primaryDefinitionOrDeclarationOccurrence(ofUSR: usr) else {
+      return nil
+    }
+    let moduleName = topLevelSymbolOccurrence.location.moduleName
+    var components = [topLevelSymbolOccurrence]
+    // Find any parent symbols
+    var symbolOccurrence: SymbolOccurrence = topLevelSymbolOccurrence
+    while let parentSymbolOccurrence = symbolOccurrence.parent(index) {
+      components.insert(parentSymbolOccurrence, at: 0)
+      symbolOccurrence = parentSymbolOccurrence
+    }
+    self.components =
+      [(name: moduleName, LinkCompletionTools.SymbolInformation(fromModuleName: moduleName))]
+      + components.map {
+        (name: $0.symbol.name, information: LinkCompletionTools.SymbolInformation(fromSymbolOccurrence: $0))
+      }
+  }
+
+  package func matches(_ link: DocCSymbolLink) -> Bool {
+    var linkComponents = link.components
+    var symbolComponents = components
+    while !linkComponents.isEmpty, !symbolComponents.isEmpty {
+      let nextLinkComponent = linkComponents.removeLast()
+      let nextSymbolComponent = symbolComponents.removeLast()
+      guard nextLinkComponent.name == nextSymbolComponent.name,
+        nextSymbolComponent.information.matches(nextLinkComponent.disambiguation)
+      else {
+        return false
+      }
+    }
+    return true
```

Comment:
This method returns `true` if `self` is a prefix of `link` or the other way round, right? I think we need to check that `linkComponets` and `symbolComponents` have the same size above. And then you should also be able to iterate over them using `zip`a

```suggestion
    guard link.components.count == components.count else {
      return false
    }
    return zip(link.components, components).allSatisfy { linkComponent, symbolComponent in
      linkComponent == symbolComponent &&
        symbolComponent.information.matches(linkComponent.disambiguation)
    }
```

Reply:
Done.